plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetColumnIOConverter.java

```diff
@@ -71,26 +76,87 @@ private ParquetColumnIOConverter() {}
         }
         if (type instanceof MapType) {
             GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
-            MapType mapType = (MapType) type;
-            GroupColumnIO keyValueColumnIO = getMapKeyValueColumn(groupColumnIO);
-            if (keyValueColumnIO.getChildrenCount() != 2) {
+            Optional<Field> keyField = getMapKeyField(context, groupColumnIO);
+            Optional<Field> valueField = getMapValueField(context, groupColumnIO);
+            if (keyField.isEmpty() || valueField.isEmpty()) {
                 return Optional.empty();
             }
-            Optional<Field> keyField = constructField(mapType.getKeyType(), keyValueColumnIO.getChild(0));
-            Optional<Field> valueField = constructField(mapType.getValueType(), keyValueColumnIO.getChild(1));
             return Optional.of(new GroupField(type, repetitionLevel, definitionLevel, required, ImmutableList.of(keyField, valueField)));
         }
         if (type instanceof ArrayType) {
             GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
-            List<Type> types = type.getTypeParameters();
-            if (groupColumnIO.getChildrenCount() != 1) {
+            Optional<Field> field = getArrayElementField(context, groupColumnIO);
+            if (field.isEmpty()) {
```

Comment:
Note that this refactor isn't absolutely side-effect free. 
Previously, we would bail out when `groupColumnIO` has child count != 1.
And then if `constructField` returned empty, we would return `new GroupField` with single empty element.

After the change, we return empty in both cases.

The first attempt at the refactor (the one with more abstract methods, see https://github.com/trinodb/trino/pull/9104#discussion_r701950561) avoided this situation, but wasn't pretty. Alternatively, https://github.com/trinodb/trino/pull/9124 avoids this too, by splitting `ParquetColumnIOConverter` instead of reusing it.